usr/src/uts/common/fs/zfs/zap_leaf.c

```diff
@@ -361,7 +362,7 @@ zap_leaf_array_match(zap_leaf_t *l, zap_name_t *zn,
 	}
 
 	ASSERT(zn->zn_key_intlen == 1);
-	if (zn->zn_matchtype == MT_FIRST) {
+	if (zn->zn_matchtype & MT_FORM) {
```

Comment:
MT_CASE isn't about ignoring the case, it's about honoring (preserving) the case in lookup request, so that we reply with what we were asked about if there's a match.

All normalization settings (both unicode and case-folding) are specified in other flags, MT_FORM specifies whether we should use the flags specified in z_norm, MT_CASE is about honoring the lookup case.